extensions/godslayerakp/http.js

```diff
@@ -0,0 +1,445 @@
+        }
+
+        getHeaderValue(args) {
+            const name = Cast.toString(args.name);
+            return this.response.get(name);
+        }
+
+        getHeaderJSON() {
+            const object = {};
+            for (const entry of this.response.headers.entries()) {
+                object[entry[0]] = entry[1];
+            }
+            return JSON.stringify(object);
+        }
+
+        /* -------- EVENTS -------- */
+
+        onResponse() {
+            const { events } = this.request;
+            return events.check('reqEnd');
+        }
+
+        onFail() {
+            const { events } = this.request;
+            return events.check('reqFail');
+        }
+
+        /* -------- CONTROL --------- */
+
+        setMimeType(args) {
+            const type = Cast.toString(args.type);
+            this.request.mimeType = type;
+        }
+
+        setRequestmethod(args) {
+            const method = Cast.toString(args.method);
+            this.request.method = method;
+        }
+
+        setHeaderData(args) {
+            const key = Cast.toString(args.name);
+            const value = Cast.toString(args.value);
+            this.request.options.headers[key] = value;
+        }
+
+        setHeaderJSON(args) {
+            const json = Cast.toString(args.json);
+            let object;
+            // ignore invalid data
+            try {
+                object = JSON.parse(json);
+            } catch {
+                return;
+            }
+            if (typeof object !== 'object') return;
+            this.request.options.headers = object;
+        }
+
+        setBody(args) {
+            const body = Cast.toString(args.text);
+            this.request.body = body;
+        }
+
+        // eslint-disable-next-line require-await
+        async sendRequest(args) {
+            const url = Cast.toString(args.url);
+            const {request, response} = this;
+
+            response.url = url;
+            // @ts-ignore
+            Scratch.fetch(url, request.options)
+                .then(res => {
+                    // @ts-ignore
+                    response.status = res.status;
+                    response.headers = res.headers;
+                    response.statusText = res.statusText;
+                    request.events.activate(res.ok ? 'reqSuccess' : 'reqFail');
+                    request.events.activate('reqEnd');
+                    return res.text();
+                })
+                .then(body => response.text = body)
+                .catch(err => {
+                    response.error = String(err);
+                    console.warn('request failed with error', err);
+                    request.events.activate('reqFail');
+                    request.events.activate('reqEnd');
+                });
+        }
+    }
+
+    const instance = new WebRequests();
+    Scratch.extensions.register(instance);
+    // @ts-ignore
+    runtime.ext_http = instance;

```

Comment:
you shouldn't need to do that

Reply:
is there a good reason those extra blocks can't just be in the extension? there is no precedent for an extension to require another one to be used first

Reply:
> is there a good reason those extra blocks can't just be in the extension? there is no precedent for an extension to require another one to be used first

yes and no, i put them separately becuase they look ugly and dont rlly fit well into the main extension imo

Reply:
reading it again i noticed you specified "required", they are not meant to be required and for the most part are not

Reply:
it seems to me that every block in httpExtra uses ext_http, so yes it is a required dependency

Reply:
> it seems to me that every block in httpExtra uses ext_http, so yes it is a required dependency

oh you meant the other way
i was reading it as http requiring httpExtra